### PR TITLE
Fix issue #384 (invisible legends when y is negative)

### LINKED ISSUE #384: Negative y value for legend causes error and invisible legend
If you take the code given in one of the examples (https://plot.ly/javascript/legend/#positioning-the-legend-outside-the-plot) and make the y value for the legend negative, the legend will disappear and errors will appear in the console. This makes it impossible to position the legend below the graph.

Here is a codepen with the stated modification already made:
http://codepen.io/anon/pen/eZGWXo

### PR DESCRIPTION
**Pull request by n-riesco**
This a new solution to fix #384 that only touches `Legend.draw()`

### CODE REVIEW

File: src/components/legend/draw.js [hunk 1]
```diff
@@ -184,27 +184,51 @@ module.exports = function draw(gd) {
...
         ly -= opts.height;
     }
-    if(anchorUtils.isMiddleAnchor(opts)) {
+    else if(anchorUtils.isMiddleAnchor(opts)) {
         ly -= opts.height / 2;
     }
 
+    //lx = Math.round(lx);
+    //ly = Math.round(ly);
+
```
Comment (n-riesco) [inline src/components/legend/draw.js]: I've commented out these two lines only to reproduce the baseline images, but I think they should be brought back.

File: src/components/legend/draw.js [hunk 2]
```diff
@@ -184,27 +184,51 @@ module.exports = function draw(gd) {
...
+        if(ly < lyMin) ly = lyMin;
+
+        legendHeight = Math.min(lyMax - ly, opts.height);
+    }
+
     // Deal with scrolling
-    var plotHeight = fullLayout.height - fullLayout.margin.b,
-        scrollheight = Math.min(plotHeight - ly, opts.height),
-        scrollPosition = scrollBox.attr('data-scroll') ? scrollBox.attr('data-scroll') : 0;
+    var scrollPosition = scrollBox.attr('data-scroll') ?
```
Comment (mdtusz) [inline src/components/legend/draw.js]: For conditional assignments like this, we can make it a bit more concise/readable:

``` js
var scrollPosition = scrollBox.attr('data-scroll') || 0;
```

and use ternary operators in situations where the conditional value isn''t used except as the condition.

File: src/components/legend/draw.js [hunk 3]
```diff
@@ -184,27 +184,51 @@ module.exports = function draw(gd) {
...
         ly -= opts.height / 2;
     }
 
+    //lx = Math.round(lx);
+    //ly = Math.round(ly);
+
+    // Make sure the legend top and bottom are visible
+    // (legends with a scroll bar are not allowed to stretch beyond the extended
+    // margins)
+    var lyMin = 0;
```
Comment (mdtusz) [inline src/components/legend/draw.js]: 🐄  comma separated var declarations e.g.

```
var lyMin = 0,
    lyMax = fullLayout.margin.t + fu...
```

File: src/components/legend/draw.js [hunk 4]
```diff
@@ -184,27 +184,51 @@ module.exports = function draw(gd) {
...
+    var lyMax = fullLayout.margin.t + fullLayout.height + fullLayout.margin.b;
+    var legendHeight = opts.height;
+    var legendHeightMax = gs.h;
```
Comment (mdtusz) [inline src/components/legend/draw.js]: We're only using the `legendHeightMax` in the one condition below, so no need to define a new var here - just adds more for the GC to eat later.

File: src/components/legend/draw.js [hunk 5]
```diff
@@ -184,27 +184,51 @@ module.exports = function draw(gd) {
...
     }
 
+    //lx = Math.round(lx);
+    //ly = Math.round(ly);
+
+    // Make sure the legend top and bottom are visible
+    // (legends with a scroll bar are not allowed to stretch beyond the extended
+    // margins)
+    var lyMin = 0;
+    var lyMax = fullLayout.margin.t + fullLayout.height + fullLayout.margin.b;
```
Comment (mdtusz) [inline src/components/legend/draw.js]: Why are we setting this to the full height + margins? Shouldn't it just be `fullLayout.height`?

**Comment by mdtusz** [comment]
This looks pretty good so far - could you elaborate on the commented out `lx` `ly` for the image baselines?

I'm going to continue testing to see if there's any issues but I haven't found anything yet.

File: src/components/legend/draw.js (hunk 4 above)
Comment (n-riesco) [inline src/components/legend/draw.js]: I'm using the variable names `lyMin`, `lyMax`, `legendHeight` and `legendHeightMax` for documentation purposes (implicitly, to describe the algorithm to compute `legendHeight`).

File: src/components/legend/draw.js (hunk 5 above)
Comment (n-riesco) [inline src/components/legend/draw.js]: Oh! Good catch!

**Comment by n-riesco** [comment]
The baseline images have been generated using `lx` and `ly` without rounding off.

When `lx` and `ly` are rounded off, many of the image tests (~75) fail with differences that are only 1 pixel.

Before introducing scrollbars, `lx` and `ly` were rounded off, and rounding off `lx` and `ly` makes it easier to reproduce plots in machines with different precisions.

**Comment by n-riesco** [comment]
* I noticed that when I drag the legend scrollbar, the mouse would move faster. I also noticed using the wheel would hardly scroll the legend. With commit 699b496 I've fixed both issues. This commit requires updating some of the tests.

* I've also noticed that this PR makes the margins of one of my mocks larger. I need to investigate this further.

**Comment by n-riesco** [comment]
Commit 8baac47 adds a nice tweak to legends. Previously, the margins were expanded using the legend size, no matter whether the legend would fit or not. But since legend that don't fit are constrained to the plot box, there is no need to expand the vertical margin and make the plot box unnecessarily smaller.

Before:
before

After:
after

**Comment by n-riesco** [comment]
@mdtusz @etpinard I had to revert my last commit because of merge conflicts. I'd like to merge this PR before trying to rebase my last commit. Is that OK?

**Comment by mdtusz** [comment]
Looks great! Let's either remove [these](https://github.com/plotly/plotly.js/pull/417#discussion-diff-59258410R198) entirely or uncomment them (it seems like they may not be necessary?), then 💃  and we can merge it in!

**Comment by n-riesco** [comment]
@mdtusz Done.

I have a commit that splits `Legend.draw()` into smaller functions. It conflicts with master. Shall I try to rebase it?

While playing with the scrollbar I noticed that the top and bottom traces in the legend may overlap with the legend border. I've thought of two possible solutions:
1. move the clippath from `legend` to `scrollbox` (this implies updating the clippath position along with that of the scrollbox on drag and wheel events),
2. apply the clippath to a new `<g>` and make the scrollbox a child of this new `<g>`.

Option 2 involves less coding (but adds one more layer). Which option do you prefer?

**Comment by mdtusz** [comment]
I'd prefer to keep the layers as minimal as we can - can we just alter the clippath to be "padded" - i.e. reduce it's height by *x* pixels and translate down by *x/2* pixels?

Or perhaps I'm misunderstanding the issue.

**Comment by n-riesco** [comment]
This is a screenshot showing the problem:

border

The clippath is applied to `legend` and `legend` contains both `rect.bg` and `scrollbox`. If we shrink the clippath, then both `rect.bg` and `scrollbox` will be affected.

**Comment by mdtusz** [comment]
Ah got it. Ok, well I'd _prefer_ if we went with option 1, but I'll leave it to your judgement.

**Comment by mdtusz** [comment]
Thanks Nico. This looks good to go.

💃

**Pull request merged**